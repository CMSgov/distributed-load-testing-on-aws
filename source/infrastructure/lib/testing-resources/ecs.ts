// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnResource, Tags, Fn } from "aws-cdk-lib";
import { ContainerImage, FargateTaskDefinition, LogDriver, CfnCluster } from "aws-cdk-lib/aws-ecs";
import { Effect, ServicePrincipal, Role, PolicyDocument, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { CfnSecurityGroup, CfnSecurityGroupEgress, CfnSecurityGroupIngress } from "aws-cdk-lib/aws-ec2";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { DockerImageAsset } from "aws-cdk-lib/aws-ecr-assets";
import * as path from "path";
import { addCfnGuardSuppression } from "../common-resources/add-cfn-guard-suppression";

export interface ECSResourcesConstructProps {
  readonly fargateVpcId: string;
  readonly scenariosS3Bucket: string;
  readonly securityGroupEgress: string;
  readonly solutionId: string;
  readonly stableTagCondition: string;
}

/**
 * Distributed Load Testing on AWS Fargate and ECS test runner construct.
 * This creates the ECS cluster, Fargate task definition, and Security Group
 */
export class ECSResourcesConstruct extends Construct {
  public taskClusterName: string;
  public ecsCloudWatchLogGroup: LogGroup;
  public taskDefinitionArn: string;
  public taskExecutionRoleArn: string;
  public ecsSecurityGroupId: string;
  public taskDefinitionFamily: string;

  constructor(scope: Construct, id: string, props: ECSResourcesConstructProps) {
    super(scope, id);

    const dltTaskCluster = new CfnCluster(this, "DLTEcsCluster", {
      clusterName: Aws.STACK_NAME,
      clusterSettings: [{ name: "containerInsights", value: "enabled" }],
      tags: [
        {
          key: "SolutionId",
          value: props.solutionId,
        },
        {
          key: "CloudFormation Stack",
          value: Aws.STACK_NAME,
        },
      ],
    });

    this.taskClusterName = dltTaskCluster.ref;

    const scenariosBucketArn = Bucket.fromBucketName(this, "ScenariosBucket", props.scenariosS3Bucket).bucketArn;

    const dltTaskExecutionRole = new Role(this, "DLTTaskExecutionRole", {
      assumedBy: new ServicePrincipal("ecs-tasks.amazonaws.com"),
      inlinePolicies: {
        ECSTaskExecutionPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "ecr:GetAuthorizationToken",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              resources: ["*"],
            }),
          ],
        }),
        ScenariosS3Policy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:ListBucket"],
              resources: [scenariosBucketArn, `${scenariosBucketArn}/*`],
            }),
          ],
        }),
      },
    });
    this.taskExecutionRoleArn = dltTaskExecutionRole.roleArn;

    addCfnGuardSuppression(dltTaskExecutionRole, "IAM_POLICYDOCUMENT_NO_WILDCARD_RESOURCE");
    addCfnGuardSuppression(dltTaskExecutionRole, "IAM_NO_INLINE_POLICY_CHECK");

    this.ecsCloudWatchLogGroup = new LogGroup(this, "DLTCloudWatchLogsGroup", {
      retention: RetentionDays.ONE_YEAR,
    });
    const dltLogsGroupResource = this.ecsCloudWatchLogGroup.node.defaultChild as CfnResource;
    dltLogsGroupResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });
    Tags.of(this.ecsCloudWatchLogGroup).add("SolutionId", props.solutionId);

    const dltTaskDefinition = new FargateTaskDefinition(this, "DLTTaskDefinition", {
      cpu: 2048,
      memoryLimitMiB: 4096,
      executionRole: dltTaskExecutionRole,
      taskRole: dltTaskExecutionRole,
    });
    const dockerRepoName = "distributed-load-testing-on-aws-load-tester";

    const versionTagForImage: string =
      process.env.PUBLIC_ECR_REGISTRY && process.env.PUBLIC_ECR_TAG
        ? `${process.env.PUBLIC_ECR_REGISTRY}/${dockerRepoName}:${process.env.PUBLIC_ECR_TAG}`
        : "";

    const stageTagForImage: string =
      process.env.PUBLIC_ECR_REGISTRY && process.env.PUBLIC_ECR_TAG
        ? `${process.env.PUBLIC_ECR_REGISTRY}/${dockerRepoName}:${
            process.env.PUBLIC_ECR_TAG.substring(0, 4) + "_stable"
          }`
        : "";

    const imageUseTag = Fn.conditionIf(props.stableTagCondition, stageTagForImage, versionTagForImage).toString();

    const imageAsset = imageUseTag
      ? ContainerImage.fromRegistry(`${imageUseTag}`)
      : new DockerImageAsset(this, "LoadTesterImage", {
          directory: path.join(__dirname, `../../../../deployment/ecr/${dockerRepoName}`),
        });

    dltTaskDefinition.addContainer("LoadTestContainer", {
      containerName: `${Aws.STACK_NAME}-load-tester`,
      image: imageAsset instanceof DockerImageAsset ? ContainerImage.fromDockerImageAsset(imageAsset) : imageAsset,
      memoryLimitMiB: 4096,
      logging: LogDriver.awsLogs({
        streamPrefix: "load-testing",
        logGroup: this.ecsCloudWatchLogGroup,
      }),
    });
    Tags.of(dltTaskDefinition).add("SolutionId", props.solutionId);
    this.taskDefinitionArn = dltTaskDefinition.taskDefinitionArn;

    const ecsSecurityGroup = new CfnSecurityGroup(this, "DLTEcsSecurityGroup", {
      vpcId: props.fargateVpcId,
      groupDescription: "DLTS Tasks Security Group",
    });
    ecsSecurityGroup.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W40",
          reason: "IpProtocol set to -1 (any) as ports are not known prior to running tests",
        },
        {
          id: "F1000",
          reason: "egress rule is specificed as its own cfn resource.",
        },
      ],
    });

    this.ecsSecurityGroupId = ecsSecurityGroup.ref;

    new CfnSecurityGroupEgress(this, "DLTSecGroupEgress", {
      cidrIp: props.securityGroupEgress,
      description: "Allow tasks to call out to external resources",
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "-1",
    });

    new CfnSecurityGroupIngress(this, "DLTSecGroupIngress", {
      description: "Allow tasks to communicate",
      fromPort: 50000,
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "tcp",
      sourceSecurityGroupId: ecsSecurityGroup.ref,
      toPort: 50000,
    });

    this.taskDefinitionFamily = dltTaskDefinition.family;
  }
}
