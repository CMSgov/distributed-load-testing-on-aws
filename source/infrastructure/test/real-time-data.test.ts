// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { LogGroup } from 'aws-cdk-lib/aws-logs';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { RealTimeDataConstruct } from '../lib/testing-resources/real-time-data';

test('DLT real time data resources Test', () => {
  const stack = new Stack();
  const testPolicy = new Policy(stack, 'TestPolicy', {
    statements: [
      new PolicyStatement({
        resources: ['*'],
        actions: ['cloudwatch:Get*']
      })
    ]
  });
  const testLogGroup = new LogGroup(stack, 'TestLogsGroup');
  const testBucket = Bucket.fromBucketName(stack, 'SourceCodeBucket', 'testbucket');

  const realTimeData = new RealTimeDataConstruct(stack, 'TestECS', {
    cloudWatchLogsPolicy: testPolicy,
    ecsCloudWatchLogGroup: testLogGroup,
    iotEndpoint: 'iotEndpoint',
    mainRegion: 'test-region-1',
    solutionId: 'testID',
    solutionVersion: 'testVersion',
    sourceCodeBucket: testBucket,
    sourceCodePrefix: 'testPrefix'
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(stack).toHaveResourceLike("AWS::Lambda::Function", {
    Code: {
      S3Bucket: "testbucket",
      S3Key: "testPrefix/real-time-data-publisher.zip",
    },
    Environment: {
      Variables: {
        MAIN_REGION: 'test-region-1',
        IOT_ENDPOINT: 'iotEndpoint',
        SOLUTION_ID: "testID",
        VERSION: "testVersion",
      },
    },
    Handler: "index.handler",
    Runtime: "nodejs14.x",
    Timeout: 180,
  });
  expect(stack).toHaveResourceLike("AWS::Logs::SubscriptionFilter", {
    "FilterPattern": '"INFO: Current:" "live=true"',
  });
  expect(stack).toHaveResourceLike("AWS::IAM::Role", {
    "Policies": [
      {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:Publish",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:test-region-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":topic/*",
                  ],
                ],
              },
            },
          ]
        }
      }
    ]
  });
});