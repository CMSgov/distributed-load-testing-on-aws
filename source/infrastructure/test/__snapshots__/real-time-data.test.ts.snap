// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT real time data resources Test 1`] = `
Object {
  "Resources": Object {
    "TestECSRealTimeDataPublisherF2114D5B": Object {
      "DependsOn": Array [
        "TestECSrealTimeDataPublisherRoleEFE9F1CD",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            Object {
              "id": "W89",
              "reason": "This Lambda function does not require a VPC",
            },
            Object {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "testbucket",
          "S3Key": "testPrefix/real-time-data-publisher.zip",
        },
        "Description": "Real time data publisher",
        "Environment": Object {
          "Variables": Object {
            "IOT_ENDPOINT": "iotEndpoint",
            "MAIN_REGION": "test-region-1",
            "SOLUTION_ID": "testID",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestECSrealTimeDataPublisherRoleEFE9F1CD",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestECSrealTimeDataPublisherRoleEFE9F1CD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "iot:Publish",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:test-region-1:",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":topic/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "IoTPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestLogsGroup54B681C7": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestLogsGroupECSLogSubscriptionFilterBFBFAB24": Object {
      "Properties": Object {
        "DestinationArn": Object {
          "Fn::GetAtt": Array [
            "TestECSRealTimeDataPublisherF2114D5B",
            "Arn",
          ],
        },
        "FilterPattern": "\\"INFO: Current:\\" \\"live=true\\"",
        "LogGroupName": Object {
          "Ref": "TestLogsGroup54B681C7",
        },
      },
      "Type": "AWS::Logs::SubscriptionFilter",
    },
    "TestLogsGroupECSLogSubscriptionFilterCanInvokeLambdaD7381D91": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TestECSRealTimeDataPublisherF2114D5B",
            "Arn",
          ],
        },
        "Principal": "logs.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "TestLogsGroup54B681C7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestPolicyCC05E598": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": Array [
          Object {
            "Ref": "TestECSrealTimeDataPublisherRoleEFE9F1CD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
